File: experiment-13/runs/language=go_model=opus-4.8-fast_prompt=TDD/rep3/internal/mcpserver/format.go
// Context: Brazilian Soccer MCP Server.
// File: format.go
// Purpose: The tool handler implementations — each reads coerced arguments,
// runs the corresponding query, and renders a concise, human-readable answer
// in the style shown in the specification.
package mcpserver

import (
	"fmt"
	"strings"

	"brazilian-soccer-mcp/internal/soccer"
)

const defaultMatchLimit = 30

// formatMatchLine renders a single match as e.g.
// "2023-09-03: Flamengo 2-1 Fluminense (Brasileirão Round 22)".
func formatMatchLine(m soccer.Match) string {
	date := "date unknown"
	if m.HasDate {
		date = m.Date.Format("2006-01-02")
	}
	score := "vs"
	if m.HasScore {
		score = fmt.Sprintf("%d-%d", m.HomeGoals, m.AwayGoals)
	}
	ctx := m.Competition
	switch {
	case m.Round != "":
		ctx = fmt.Sprintf("%s Round %s", m.Competition, m.Round)
	case m.Stage != "":
		ctx = fmt.Sprintf("%s %s", m.Competition, m.Stage)
	}
	return fmt.Sprintf("%s: %s %s %s (%s)", date, m.HomeTeam, score, m.AwayTeam, ctx)
}

func (h *Handler) searchMatches(args map[string]any) (string, error) {
	f := soccer.MatchFilter{
		Team:        argString(args, "team"),
		Opponent:    argString(args, "opponent"),
		Competition: argString(args, "competition"),
		Season:      argInt(args, "season"),
	}
	if from := argString(args, "from"); from != "" {
		if t, ok := soccer.ParseDate(from); ok {
			f.From = t
		}
	}
	if to := argString(args, "to"); to != "" {
		if t, ok := soccer.ParseDate(to); ok {
			f.To = t
		}
	}
	limit := argInt(args, "limit")
	if limit <= 0 {
		limit = defaultMatchLimit
	}

	matches := h.DB.FindMatches(f)
	if len(matches) == 0 {
		return "No matches found for the given criteria.", nil
	}

	var b strings.Builder
	fmt.Fprintf(&b, "Found %d match(es):\n", len(matches))
	shown := matches
	if len(shown) > limit {
		shown = shown[:limit]
	}
	for _, m := range shown {
		fmt.Fprintf(&b, "- %s\n", formatMatchLine(m))
	}
	if len(matches) > limit {
		fmt.Fprintf(&b, "... (%d more not shown)\n", len(matches)-limit)
	}
	return strings.TrimRight(b.String(), "\n"), nil
}

func (h *Handler) headToHead(args map[string]any) (string, error) {
	a, err := requireString(args, "team_a")
	if err != nil {
		return "", err
	}
	bTeam, err := requireString(args, "team_b")
	if err != nil {
		return "", err
	}
	h2h := h.DB.HeadToHead(a, bTeam)
	if len(h2h.Matches) == 0 {
		return fmt.Sprintf("No matches found between %s and %s in the dataset.", a, bTeam), nil
	}
	var b strings.Builder
	fmt.Fprintf(&b, "%s vs %s — head-to-head (%d matches):\n", a, bTeam, len(h2h.Matches))
	fmt.Fprintf(&b, "%s: %d win(s), %s: %d win(s), Draws: %d\n", a, h2h.AWins, bTeam, h2h.BWins, h2h.Draws)
	fmt.Fprintf(&b, "Goals: %s %d - %d %s\n", a, h2h.AGoals, h2h.BGoals, bTeam)
	b.WriteString("\nMatches:\n")
	for _, m := range h2h.Matches {
		fmt.Fprintf(&b, "- %s\n", formatMatchLine(m))
	}
	return strings.TrimRight(b.String(), "\n"), nil
}

func parseVenue(s string) soccer.Venue {
	switch strings.ToLower(strings.TrimSpace(s)) {
	case "home":
		return soccer.VenueHome
	case "away":
		return soccer.VenueAway
	default:
		return soccer.VenueAny
	}
}

func (h *Handler) teamRecord(args map[string]any) (string, error) {
	team, err := requireString(args, "team")
	if err != nil {
		return "", err
	}
	f := soccer.TeamFilter{
		Team:        team,
		Season:      argInt(args, "season"),
		Competition: argString(args, "competition"),
		Venue:       parseVenue(argString(args, "venue")),
	}
	r := h.DB.TeamRecord(f)

	scope := team
	var quals []string
	if f.Season != 0 {
		quals = append(quals, fmt.Sprintf("%d", f.Season))
	}
	if f.Competition != "" {
		quals = append(quals, f.Competition)
	}
	switch f.Venue {
	case soccer.VenueHome:
		quals = append(quals, "home")
	case soccer.VenueAway:
		quals = append(quals, "away")
	}
	if len(quals) > 0 {
		scope = fmt.Sprintf("%s (%s)", team, strings.Join(quals, " "))
	}

	if r.Matches == 0 {
		return fmt.Sprintf("No matches found for %s.", scope), nil
	}
	var b strings.Builder
	fmt.Fprintf(&b, "%s record:\n", scope)
	fmt.Fprintf(&b, "- Matches: %d\n", r.Matches)
	fmt.Fprintf(&b, "- Wins: %d, Draws: %d, Losses: %d\n", r.Wins, r.Draws, r.Losses)
	fmt.Fprintf(&b, "- Goals For: %d, Goals Against: %d\n", r.GoalsFor, r.GoalsAgainst)
	fmt.Fprintf(&b, "- Points: %d\n", r.Points())
	fmt.Fprintf(&b, "- Win rate: %.1f%%", r.WinRate()*100)
	return b.String(), nil
}

func (h *Handler) standings(args map[string]any) (string, error) {
	season := argInt(args, "season")
	if season == 0 {
		return "", fmt.Errorf("missing required argument %q", "season")
	}
	competition := argString(args, "competition")
	if competition == "" {
		competition = soccer.CompBrasileirao
	}
	table := h.DB.Standings(season, competition)
	if len(table) == 0 {
		return fmt.Sprintf("No standings available for %d %s.", season, competition), nil
	}
	var b strings.Builder
	fmt.Fprintf(&b, "%d %s standings (calculated from matches):\n", season, competition)
	for i, r := range table {
		fmt.Fprintf(&b, "%d. %s - %d pts (%dW %dD %dL, GF %d GA %d, GD %+d)\n",
			i+1, r.Team, r.Points(), r.Wins, r.Draws, r.Losses, r.GoalsFor, r.GoalsAgainst, r.GoalDifference())
	}
	return strings.TrimRight(b.String(), "\n"), nil
}

const defaultPlayerLimit = 20

func (h *Handler) searchPlayers(args map[string]any) (string, error) {
	limit := argInt(args, "limit")
	if limit <= 0 {
		limit = defaultPlayerLimit
	}
	f := soccer.PlayerFilter{
		Name:        argString(args, "name"),
		Nationality: argString(args, "nationality"),
		Club:        argString(args, "club"),
		Position:    argString(args, "position"),
		MinOverall:  argInt(args, "min_overall"),
		Limit:       limit,
	}
	players := h.DB.FindPlayers(f)
	if len(players) == 0 {
		return "No players found for the given criteria.", nil
	}
	var b strings.Builder
	fmt.Fprintf(&b, "Found %d player(s):\n", len(players))
	for i, p := range players {
		fmt.Fprintf(&b, "%d. %s - Overall: %d, Position: %s, Club: %s, Nationality: %s\n",
			i+1, p.Name, p.Overall, p.Position, p.Club, p.Nationality)
	}
	return strings.TrimRight(b.String(), "\n"), nil
}

func (h *Handler) matchStatistics(args map[string]any) (string, error) {
	f := soccer.MatchFilter{
		Team:        argString(args, "team"),
		Competition: argString(args, "competition"),
		Season:      argInt(args, "season"),
	}
	avg := h.DB.AverageGoals(f)
	homeRate := h.DB.HomeWinRate(f)
	biggest := h.DB.BiggestWins(f, 5)
	if len(biggest) == 0 && avg == 0 {
		return "No matches found for the given criteria.", nil
	}

	var scope []string
	if f.Competition != "" {
		scope = append(scope, f.Competition)
	}
	if f.Season != 0 {
		scope = append(scope, fmt.Sprintf("%d", f.Season))
	}
	if f.Team != "" {
		scope = append(scope, f.Team)
	}
	title := "Match statistics"
	if len(scope) > 0 {
		title = fmt.Sprintf("Match statistics (%s)", strings.Join(scope, " "))
	}

	var b strings.Builder
	fmt.Fprintf(&b, "%s:\n", title)
	fmt.Fprintf(&b, "- Average goals per match: %.2f\n", avg)
	fmt.Fprintf(&b, "- Home win rate: %.1f%%\n", homeRate*100)
	if len(biggest) > 0 {
		b.WriteString("- Biggest victories:\n")
		for _, m := range biggest {
			fmt.Fprintf(&b, "  - %s\n", formatMatchLine(m))
		}
	}
	return strings.TrimRight(b.String(), "\n"), nil
}

File: experiments/adrianco/experiment-46-opus5/brazil/runs/language=clojure_model=claude-opus-5_prompt=neutral/rep1/src/brazilian_soccer/format.clj
(ns brazilian-soccer.format
  "Rendering of query results as the plain text an MCP client shows to an LLM.

  CONTEXT
  -------
  MCP tools return text content, so every tool answer is rendered here.  The
  wording follows the answer formats in the specification: a match line reads

      2019-10-27: Flamengo 5-0 Grêmio (Brasileirão Série A, Round 27)

  and a team record reads

      Matches: 19 | Wins: 11, Draws: 5, Losses: 3 | GF: 28, GA: 15 | Win rate: 57.9%

  Numbers are always accompanied by what they were calculated from, because the
  dataset is a sample of Brazilian football, not the complete record - standings
  and averages are derived from the matches actually present in the CSVs."
  (:require [clojure.string :as str]
            [brazilian-soccer.data :as data]))

(defn pct [x] (format "%.1f%%" (* 100.0 (or x 0.0))))
(defn dec2 [x] (format "%.2f" (double (or x 0.0))))

(defn match-line
  "One match as a single line."
  ([m] (match-line m nil))
  ([m {:keys [show-competition? show-sources?] :or {show-competition? true}}]
   (let [score (if (:played? m)
                 (str (:home-goals m) "-" (:away-goals m))
                 "(no score in dataset)")
         ctx (cond-> []
               show-competition? (conj (:competition-name m))
               (:round m)  (conj (str "Round " (:round m)))
               (:stage m)  (conj (str/capitalize (:stage m)))
               (and (not show-competition?) (:season m)) (conj (str (:season m)))
               (:venue m)  (conj (:venue m))
               (:derby m)  (conj (str "derby: " (:derby m)))
               show-sources? (conj (str "sources: " (str/join "+" (:sources m)))))]
     (str (or (:date m) "date unknown") ": "
          (:home m) " " score " " (:away m)
          (when (seq ctx) (str " (" (str/join ", " ctx) ")"))))))

(defn match-list [matches opts]
  (if (empty? matches)
    "  (no matches)"
    (str/join "\n" (map #(str "  - " (match-line % opts)) matches))))

(defn record-line [r]
  (str "Matches: " (:matches r)
       " | Wins: " (:wins r) ", Draws: " (:draws r) ", Losses: " (:losses r)
       " | Goals For: " (:goals-for r) ", Goals Against: " (:goals-against r)
       " (" (if (neg? (:goal-difference r)) "" "+") (:goal-difference r) ")"
       " | Points: " (:points r)
       " | Win rate: " (pct (:win-rate r))))

(defn filters-line [{:keys [competition season season-from season-to date-from date-to venue]}]
  (let [parts (cond-> []
                competition (conj (data/competition-name competition))
                season      (conj (str "season " season))
                season-from (conj (str "from " season-from))
                season-to   (conj (str "to " season-to))
                date-from   (conj (str "after " date-from))
                date-to     (conj (str "before " date-to))
                (and venue (not= venue :any)) (conj (str (clojure.core/name venue) " matches only")))]
    (if (seq parts) (str/join ", " parts) "all competitions and seasons")))

;; ---------------------------------------------------------------------------
;; Tool renderings
;; ---------------------------------------------------------------------------

(defn matches-answer [{:keys [title matches total filters]}]
  (str title "\n"
       (when filters (str "Filter: " filters "\n"))
       (match-list matches nil)
       "\n\n" (count matches) " shown"
       (when (and total (> total (count matches))) (str " of " total " in the dataset"))
       "."))

(defn team-stats-answer [{:keys [team overall home away by-competition by-season
                                 biggest-win recent filters]}]
  (str (:display team) (when (:state team) (str " (" (:state team) ")"))
       " - " filters "\n"
       (record-line overall) "\n"
       "  Home: " (record-line home) "\n"
       "  Away: " (record-line away) "\n"
       (when (> (count by-competition) 1)
         (str "\nBy competition:\n"
              (str/join "\n" (for [[c r] by-competition]
                               (str "  - " (data/competition-name c) ": "
                                    (:matches r) " matches, " (:wins r) "W " (:draws r) "D "
                                    (:losses r) "L, " (:goals-for r) ":" (:goals-against r))))
              "\n"))
       (when (> (count by-season) 1)
         (str "\nBy season:\n"
              (str/join "\n" (for [[s r] by-season]
                               (str "  - " s ": " (:matches r) " matches, "
                                    (:wins r) "W " (:draws r) "D " (:losses r) "L, "
                                    (:points r) " pts")))
              "\n"))
       (when biggest-win (str "\nBiggest win: " (match-line biggest-win) "\n"))
       (when (seq recent)
         (str "\nMost recent matches:\n" (match-list recent nil)))))

(defn head-to-head-answer [{:keys [team-a team-b matches played a-record b-record draws
                                   by-competition first-meeting last-meeting biggest]}
                           {:keys [limit filters]}]
  (let [a (:display team-a) b (:display team-b)]
    (if (zero? played)
      (str "No matches between " a " and " b " in the dataset (" filters ").")
      (str a " vs " b " - head-to-head (" filters ")\n"
           "Matches: " played
           " | " a " wins: " (:wins a-record)
           " | " b " wins: " (:wins b-record)
           " | Draws: " draws "\n"
           "Goals: " a " " (:goals-for a-record) " - " (:goals-for b-record) " " b "\n"
           "First meeting: " (match-line first-meeting) "\n"
           "Last meeting:  " (match-line last-meeting) "\n"
           (when biggest (str "Biggest margin: " (match-line biggest) "\n"))
           "\nBy competition:\n"
           (str/join "\n" (for [[c agg] by-competition
                                :let [ra (:a agg) rb (:b agg)]]
                            (str "  - " (data/competition-name c) ": " (:matches agg)
                                 (if (= 1 (:matches agg)) " match, " " matches, ")
                                 a " " (:wins ra) "W - " (:wins rb) "W " b
                                 ", " (:draws ra) " draws")))
           "\n\nMatches (most recent first):\n"
           (match-list (take (or limit 20) matches) nil)
           (when (> (count matches) (or limit 20))
             (str "\n  ... " (- (count matches) (or limit 20)) " more"))))))

(defn standings-answer [{:keys [competition season table match-count]}]
  (let [teams (count table)
        relegation (when (and (= :serie-a competition) (>= teams 20)) 4)]
    (str (data/competition-name competition) " " season
         " - table calculated from " match-count " matches in the dataset\n"
         (format "%-4s %-26s %3s %3s %3s %3s %4s %4s %4s %5s"
                 "#" "Team" "P" "W" "D" "L" "GF" "GA" "GD" "Pts") "\n"
         (str/join "\n"
                   (for [r table]
                     (str (format "%-4s %-26s %3s %3s %3s %3s %4s %4s %4s %5s"
                                  (str (:position r)) (:team r) (:matches r) (:wins r)
                                  (:draws r) (:losses r) (:goals-for r) (:goals-against r)
                                  (str (when (pos? (:goal-difference r)) "+") (:goal-difference r))
                                  (:points r))
                          (cond
                            (= 1 (:position r)) "  <- champion (by points)"
                            (and relegation (> (:position r) (- teams relegation))) "  <- relegation zone"
                            :else ""))))
         "\n\nCalculated from match results only: 3 points per win, ranked on"
         " points, wins, goal difference, goals for.")))

(defn rankings-answer [{:keys [metric venue min-matches rows]} filters]
  (str "Clubs ranked by " (clojure.core/name metric)
       (when-not (= :any venue) (str " (" (clojure.core/name venue) " matches only)"))
       " - " filters "\n"
       "Minimum " min-matches " matches played.\n"
       (str/join "\n"
                 (for [r rows]
                   (str "  " (:position r) ". " (:team r) " - "
                        (case metric
                          :points-per-match (str (dec2 (:points-per-match r)) " pts/match")
                          :win-rate (str (pct (:win-rate r)) " win rate")
                          :goals-per-match (str (dec2 (:goals-per-match r)) " goals/match")
                          (str (get r metric) " " (clojure.core/name metric)))
                        " (" (record-line r) ")")))))

(defn competition-summary-answer [{:keys [competition season matches goals goals-per-match
                                          home-wins away-wins draws home-win-rate draw-rate
                                          away-win-rate seasons teams biggest-win
                                          highest-scoring top-scoring-teams]}]
  (str (if competition (data/competition-name competition) "All competitions")
       (when season (str " " season)) " - statistics from the dataset\n"
       "Matches: " matches
       (when (and (not season) (seq seasons))
         (str " across seasons " (first seasons) "-" (last seasons)))
       "\n"
       "Distinct teams: " teams "\n"
       "Goals: " goals " (" (dec2 goals-per-match) " per match)\n"
       "Home wins: " home-wins " (" (pct home-win-rate) ")"
       " | Away wins: " away-wins " (" (pct away-win-rate) ")"
       " | Draws: " draws " (" (pct draw-rate) ")\n"
       (when biggest-win (str "Biggest margin: " (match-line biggest-win) "\n"))
       (when highest-scoring (str "Highest scoring: " (match-line highest-scoring) "\n"))
       (when (seq top-scoring-teams)
         (str "\nTop scoring teams:\n"
              (str/join "\n" (map-indexed (fn [i t] (str "  " (inc i) ". " (:team t) " - " (:goals t) " goals"))
                                          top-scoring-teams))))
       "\n\nNote: goalscorer data is not part of any provided dataset, so individual"
       " top scorers cannot be derived; team totals are shown instead."))

(defn player-line [p]
  (str (:name p)
       " - Overall: " (:overall p)
       ", Potential: " (:potential p)
       ", Position: " (or (:position p) "n/a")
       ", Age: " (:age p)
       ", Club: " (or (:club p) "free agent")
       ", Nationality: " (:nationality p)))

(defn players-answer [{:keys [title players total note]}]
  (str title "\n"
       (if (empty? players)
         "  (no players matched)"
         (str/join "\n" (map-indexed (fn [i p] (str "  " (inc i) ". " (player-line p))) players)))
       "\n\n" (count players) " shown of " total " matching players."
       (when note (str "\n" note))))

(defn player-profile-answer [{:keys [player team-id alternatives]} team]
  (let [p player]
    (str (:name p) " - FIFA 19 player profile\n"
         "Nationality: " (:nationality p) " | Age: " (:age p)
         " | Position: " (or (:position p) "n/a")
         (when (:jersey p) (str " | Shirt: " (:jersey p))) "\n"
         "Club: " (or (:club p) "free agent")
         (when team-id (str " (linked to " (:display team) " in the match data)")) "\n"
         "Overall: " (:overall p) " | Potential: " (:potential p)
         " | Value: " (or (:value p) "n/a") " | Wage: " (or (:wage p) "n/a") "\n"
         "Height: " (or (:height p) "n/a") " | Weight: " (or (:weight p) "n/a")
         " | Preferred foot: " (or (:foot p) "n/a") "\n"
         (when (seq (:skills p))
           (str "Top attributes: "
                (->> (:skills p) (sort-by (comp - val)) (take 8)
                     (map (fn [[k v]] (str k " " v)))
                     (str/join ", "))
                "\n"))
         (when (seq alternatives)
           (str "\nOther players matching that name: "
                (str/join ", " (map #(str (:name %) " (" (:club %) ", " (:overall %) ")")
                                    alternatives)))))))

(defn squad-answer [{:keys [clubs players player-count brazilians avg-overall nationalities]}
                    team]
  (str "Squad: " (str/join " / " clubs)
       (when team (str " - " (:display team) " in the match data")) "\n"
       "Players: " player-count " | Brazilians: " brazilians
       " | Average overall rating: " (dec2 avg-overall) "\n"
       "Nationalities: " (str/join ", " (map (fn [[n c]] (str n " " c)) nationalities)) "\n\n"
       (str/join "\n" (map-indexed (fn [i p] (str "  " (inc i) ". " (player-line p)))
                                   (take 30 players)))))

(defn teams-answer [teams]
  (str "Clubs in the dataset (" (count teams) " shown)\n"
       (str/join "\n"
                 (for [t teams]
                   (str "  - " (:display t)
                        (when (:state t) (str " [" (:state t) "]"))
                        (when (:country t) (str " [" (:country t) "]"))
                        " - " (:match-count t) " matches, competitions: "
                        (str/join ", " (map data/competition-name (:competitions t)))
                        "\n    id: " (:id t)
                        " | spellings in the data: " (str/join " / " (:variants t)))))))

(defn dataset-answer [{:keys [data-dir sources stats matches played teams players
                              linked-clubs by-competition date-range]}]
  (str "Brazilian soccer knowledge graph\n"
       "Data directory: " data-dir "\n"
       "Rows read: " (:raw-match-rows stats)
       " -> unique matches after de-duplicating overlapping files: " matches
       " (" played " with a score)\n"
       "Clubs: " teams " | FIFA players: " players
       " | FIFA squads linked to clubs in the match data: " linked-clubs "\n"
       "Match dates: " (first date-range) " to " (second date-range) "\n\n"
       "Competitions:\n"
       (str/join "\n" (for [c by-competition]
                        (str "  - " (:name c) ": " (:matches c) " matches, seasons "
                             (first (:seasons c)) "-" (second (:seasons c)))))
       "\n\nSource files:\n"
       (str/join "\n" (for [s sources]
                        (str "  - " (:file s) " - " (:title s)
                             "\n    license: " (:license s) " | " (:url s))))
       "\n\nRows per file: "
       (str/join ", " (for [[k v] (sort-by key (:rows-by-source stats))]
                        (str (clojure.core/name k) " " v)))
       "\nMatches appearing in more than one file are merged, so a fixture is"
       " counted once while keeping the round, stadium and shot/corner detail"
       " each file contributes."))
